src/acl/ParameterizedNode.h

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright (C) 1996-2023 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+#ifndef SQUID_SRC_ACL_PARAMETERIZEDNODE_H
+#define SQUID_SRC_ACL_PARAMETERIZEDNODE_H
+
+#include "acl/Acl.h"
+#include "acl/Data.h"
+
+#include <memory>
+
+namespace Acl
+{
+
+/// An ACL that manages configured parameters using a given Parameters class P.
+/// That Parameters class must support an ACLData<> or equivalent API.
+template <class P>
+class ParameterizedNode: public ACL
+{
+public:
+    using Parameters = P;
+
+    // to avoid dragging constructor parameters through each derived class, they
+    // are set in a leaf class constructor; \sa Acl::FinalizedParameterizedNode
+    ParameterizedNode() = default;
+    ~ParameterizedNode() override = default;
+
+protected:
+    /* ACL API */
+    void prepareForUse() override { parameters->prepareForUse(); }
+    void parse() override { parameters->parse(); }
+    SBufList dump() const override { return parameters->dump(); }
+    bool empty() const override { return parameters->empty(); }
+    const Acl::Options &lineOptions() override { return parameters->lineOptions(); }
+
+    /// items this ACL is configured to match; never nil after construction
+    std::unique_ptr<Parameters> parameters;
```

Comment:
What name do you prefer for this above std::unique_ptr<Parameters> class member (called `parameters` in this PR code), keeping in mind that this class member lives inside a ParameterizedNode class that lives inside an Acl namespace:

* `parameters` -- matches the meaning of ACLData objects it stores; we do not have consistent terminology for this, but those objects are known as ACL parameters (among other, more problematic names);

* `data` -- matches the type name of ACLData objects it stores and minimizes PR code changes, but tells nothing to the code reader (nearly everything can be called "data");

* `options` -- please do not propose "options" because that word is already used for ACL-global flags/options and ACL line parameter flags/options (see ACL::options() and ACL::lineOptions());

* `arguments` -- please do not propose "arguments" because that word is already used to describe "ACL parameter values", "ACL-global options", and "ACL line options" collectively/together (i.e. anything that follows the `acl aclname acltype` directive prefix);

* any better options?

Reply:
The wording does "smell" a bit to me, but cannot point at specifically why right now.

Reply:
> The wording does "smell" a bit to me, but cannot point at specifically why right now.

We can try to improve the "wording" later -- it is not duplicated, so it would be easy to adjust as needed. Finalizing the field _name_ **now** is important because it is duplicated -- all derived ACLs have to use that data member by name.

Reply:
I decided to preserve the old `data` name to avoid creating otherwise unnecessary and out-of-scope code changes in this PR. We can rename this class member to `parameters` (or whatever we can agree on) later, of course.